Adding Unit 3: Deep Q-Learning and Optuna Bonus (#108)
huggingface/deep-rl-class: units/en/unit3/hands-on.mdx

# Hands-on [[hands-on]]

Now that you've studied the theory behind Deep Q-Learning, **you’re ready to train your Deep Q-Learning agent to play Atari Games**. We'll start with Space Invaders, but you'll be able to use any Atari game you want 🔥

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit4/atari-envs.gif" alt="Environments"/>

We're using the [RL-Baselines-3 Zoo integration](https://github.com/DLR-RM/rl-baselines3-zoo), a vanilla version of Deep Q-Learning with no extensions such as Double-DQN, Dueling-DQN, and Prioritized Experience Replay.

**To start the hands-on click on Open In Colab button** 👇 :

> [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)]()

Reviewer: Missing the hands on

Author: Added :heavy_check_mark:

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit4/atari-envs.gif" alt="Environments"/>

We're using the [RL-Baselines-3 Zoo integration](https://github.com/DLR-RM/rl-baselines3-zoo), a vanilla version of Deep Q-Learning with no extensions such as Double-DQN, Dueling-DQN, and Prioritized Experience Replay.

To validate this hands-on for the certification process, you need to push your trained model to the Hub and **get a result of >= 500**.

To find your result, go to the leaderboard and find your model, **the result = mean_reward - std of reward**

For more information about the certification process, check this section 👉 https://huggingface.co/deep-rl-course/en/unit0/introduction#certification-process

**To start the hands-on click on Open In Colab button** 👇 :

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/deep-rl-class/blob/master/notebooks/unit3/unit3.ipynb)

> # Unit 3: Deep Q-Learning with Atari Games 👾 using RL Baselines3 Zoo

Reviewer: This notebook is very high level compared to the previous one, so maybe it's good to set up the expectations since the beginning in a more explicit way that we're not going to implement a neural network here but use a common RL library used out there. Maybe at the end, apart from the link to the implementation, we could share some link to a tutorial for implementing Deep Q Learning from scratch?

## Set the GPU 💪
- To **accelerate the agent's training, we'll use a GPU**. To do that, go to `Runtime > Change Runtime type`

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/gpu-step1.jpg" alt="GPU Step 1">

- `Hardware Accelerator > GPU`

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/gpu-step2.jpg" alt="GPU Step 2">

## Create a virtual display 🔽

During the notebook, we'll need to generate a replay video. To do so, with colab, **we need to have a virtual screen to be able to render the environment** (and thus record the frames).

Hence the following cell will install the librairies and create and run a virtual screen 🖥

```bash
apt install python-opengl
apt install ffmpeg
apt install xvfb
pip3 install pyvirtualdisplay
```

```bash
apt-get install swig cmake freeglut3-dev
```

```bash
pip install pyglet==1.5.1
```

```python
# Virtual display
from pyvirtualdisplay import Display

virtual_display = Display(visible=0, size=(1400, 900))
virtual_display.start()
```

## Clone RL-Baselines3 Zoo Repo 📚

> You can now directly install from python package `pip install rl_zoo3` but since we want **the full installation with extra environments and dependencies** we're going to clone `RL-Baselines3-Zoo` repository and install from source.

Reviewer: ```suggestion
You could directly install from the Python package (`pip install rl_zoo3`), but since we want **the full installation with extra environments and dependencies**, we're going to clone the `RL-Baselines3-Zoo` repository and install from source.
```

```bash
git clone https://github.com/DLR-RM/rl-baselines3-zoo
```

## Install dependencies 🔽
We can now install the dependencies RL-Baselines3 Zoo needs (this can take 5min ⏲)

But we'll also install:
- `huggingface_sb3`: Additional code for Stable-baselines3 to load and upload models from the Hugging Face 🤗 Hub.

```bash
cd /content/rl-baselines3-zoo/
```

```bash
pip install -r requirements.txt
```

```bash
pip install huggingface_sb3

> ```

Reviewer: `huggingface_sb3` is already specified in `requirements.txt` so not needed

Author: Updated ✔️

## Train our Deep Q-Learning Agent to Play Space Invaders 👾

To train an agent with RL-Baselines3-Zoo, we just need to do two things:
1. We define the hyperparameters in `rl-baselines3-zoo/hyperparams/dqn.yml`

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit3/hyperparameters.png" alt="DQN Hyperparameters">

Here we see that:

> - We use the `Atari Wrapper` that preprocess the input (Frame reduction ,grayscale, stack 4 frames)

Reviewer: ```suggestion
- We use the `Atari Wrapper` that does the pre-processing (Frame reduction, grayscale, stack four frames frames),
```

> - We use `CnnPolicy`, since we use Convolutional layers to process the frames

Reviewer: ```suggestion
- We use `CnnPolicy`, since we use Convolutional layers to process the frames.
```

> - We train it for 10 million `n_timesteps`

Reviewer: ```suggestion
- We train the model for 10 million `n_timesteps`.
```

> - Memory (Experience Replay) size is 100000, aka the amount of experience steps you saved to train again your agent with.

Reviewer: ```suggestion
- Memory (Experience Replay) size is 100000, i.e. the number of experience steps you saved to train again your agent with.
```

> - Memory (Experience Replay) size is 100000, aka the amount of experience steps you saved to train again your agent with.

Reviewer: This was not super clear to me

> 💡 My advice is to **reduce the training timesteps to 1M,** which will take about 90 minutes on a P100. `!nvidia-smi` will tell you what GPU you're using. At 10 million steps, this will take about 9 hours, which could likely result in Colab timing out. I recommend running this on your local computer (or somewhere else). Just click on: `File>Download`.

Reviewer: Modifying this file tells me i cannot save a copy in drive and force me to create a copy of the unit 3 colab, which creates lots of friction. I would suggest since the beginning to create a copy

In terms of hyperparameters optimization, my advice is to focus on these 3 hyperparameters:
- `learning_rate`
- `buffer_size (Experience Memory size)`
- `batch_size`

As a good practice, you need to **check the documentation to understand what each hyperparameters does**: https://stable-baselines3.readthedocs.io/en/master/modules/dqn.html#parameters

2. We run `train.py` and save the models on `logs` folder 📁

```bash
python train.py --algo ________ --env SpaceInvadersNoFrameskip-v4  -f _________
```

#### Solution

> ```bash

Reviewer: The solution is shown in the colab, make sure to hide it

Author: ✔️

python train.py --algo dqn  --env SpaceInvadersNoFrameskip-v4 -f logs/
```

## Let's evaluate our agent 👀

> - RL-Baselines3-Zoo provides `enjoy.py` to evaluate our agent.

Reviewer: Maybe add a sentence explaining what `enjoy.py` is. It's not super clear from this point

Author: ✔️

- Let's evaluate it for 5000 timesteps 🔥

```bash
python enjoy.py  --algo dqn  --env SpaceInvadersNoFrameskip-v4  --no-render  --n-timesteps _________  --folder logs/
```

#### Solution

```bash
python enjoy.py  --algo dqn  --env SpaceInvadersNoFrameskip-v4  --no-render  --n-timesteps 5000  --folder logs/
```

## Publish our trained model on the Hub 🚀

> Now that we saw we got good results after the training, we can publish our trained model on the hub 🤗 with one line of code.

Reviewer: ```suggestion
Now that we saw we got good results after the training, we can publish our trained model on the Hub with one line of code.
```

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit3/space-invaders-model.gif" alt="Space Invaders model">

> By using `rl_zoo3.push_to_hub.py` **you evaluate, record a replay, generate a model card of your agent and push it to the hub**.

Reviewer: ```suggestion
By using `rl_zoo3.push_to_hub.py`, **you evaluate, record a replay, generate a model card of your agent, and push it to the Hub**.
```

This way:
- You can **showcase our work** 🔥
- You can **visualize your agent playing** 👀
- You can **share with the community an agent that others can use** 💾

> - You can **access a leaderboard 🏆 to see how well your agent is performing compared to your classmates** 👉  https://huggingface.co/spaces/chrisjay/Deep-Reinforcement-Learning-Leaderboard

Reviewer: Wrong link to the dashboard

> To be able to share your model with the community there are three more steps to follow:

Reviewer: ```suggestion
To be able to share your model with the community, there are three more steps to follow:
```

> 1️⃣ (If it's not already done) create an account to HF ➡ https://huggingface.co/join

Reviewer: ```suggestion
1️⃣ (If it's not already done) create an account in HF ➡ https://huggingface.co/join
```

2️⃣ Sign in and then, you need to store your authentication token from the Hugging Face website.
- Create a new token (https://huggingface.co/settings/tokens) **with write role**

<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/create-token.jpg" alt="Create HF Token">

- Copy the token
- Run the cell below and past the token

```python
from huggingface_hub import notebook_login # To log to our Hugging Face account to be able to upload models to the Hub.
notebook_login()
git config --global credential.helper store
```

If you don't want to use a Google Colab or a Jupyter Notebook, you need to use this command instead: `huggingface-cli login`

> 3️⃣ We're now ready to push our trained agent to the 🤗 Hub 🔥

Reviewer: ```suggestion
3️⃣ We're now ready to push our trained agent to the Hub 🔥
```

> Let's run push_to_hub.py file to upload our trained agent to the Hub.

Reviewer: ```suggestion
Let's run `push_to_hub.py` file to upload our trained agent to the Hub. There are two important parameters:
```

> `--repo-name `: The name of the repo
> 
> `-orga`: Your Hugging Face username

Reviewer: ```suggestion
* `--repo-name `: The name of the repo
* `-orga`: Your Hugging Face username
```

#### Solution

```bash
python -m rl_zoo3.push_to_hub  --algo dqn  --env SpaceInvadersNoFrameskip-v4  --repo-name dqn-SpaceInvadersNoFrameskip-v4  -orga ThomasSimonini  -f logs/
```

Congrats 🥳 you've just trained and uploaded your first Deep Q-Learning agent using RL-Baselines-3 Zoo. The script above should have displayed a link to a model repository such as https://huggingface.co/ThomasSimonini/dqn-SpaceInvadersNoFrameskip-v4. When you go to this link, you can:

- See a **video preview of your agent** at the right.
- Click "Files and versions" to see all the files in the repository.
- Click "Use in stable-baselines3" to get a code snippet that shows how to load the model.
- A model card (`README.md` file) which gives a description of the model and the hyperparameters you used.

Under the hood, the Hub uses git-based repositories (don't worry if you don't know what git is), which means you can update the model with new versions as you experiment and improve your agent.

**Compare the results of your agents with your classmates** using the [leaderboard](https://huggingface.co/spaces/huggingface-projects/Deep-Reinforcement-Learning-Leaderboard) 🏆

## Load a powerful trained model 🔥

> - The Stable-Baselines3 team uploaded **more than 150 trained Deep Reinforcement Learning agents on the Hub**.

Reviewer: ```suggestion
The Stable-Baselines3 team uploaded **more than 150 trained Deep Reinforcement Learning agents on the Hub**. You can download them and use them to see how they perform!
```

You can find them here: 👉 https://huggingface.co/sb3

Some examples:
- Asteroids: https://huggingface.co/sb3/dqn-AsteroidsNoFrameskip-v4
- Beam Rider: https://huggingface.co/sb3/dqn-BeamRiderNoFrameskip-v4
- Breakout: https://huggingface.co/sb3/dqn-BreakoutNoFrameskip-v4
- Road Runner: https://huggingface.co/sb3/dqn-RoadRunnerNoFrameskip-v4

Let's load an agent playing Beam Rider: https://huggingface.co/sb3/dqn-BeamRiderNoFrameskip-v4

```python
<video controls autoplay><source src="https://huggingface.co/sb3/dqn-BeamRiderNoFrameskip-v4/resolve/main/replay.mp4" type="video/mp4"></video>
```

1. We download the model using `rl_zoo3.load_from_hub`, and place it in a new folder that we can call `rl_trained`

```bash
# Download model and save it into the logs/ folder
python -m rl_zoo3.load_from_hub --algo dqn --env BeamRiderNoFrameskip-v4 -orga sb3 -f rl_trained/
```

2. Let's evaluate if for 5000 timesteps

```bash
python enjoy.py --algo dqn --env BeamRiderNoFrameskip-v4 -n 5000  -f rl_trained/
```

Why not trying to train your own **Deep Q-Learning Agent playing BeamRiderNoFrameskip-v4? 🏆.**

> If you want to try, check https://huggingface.co/sb3/dqn-BeamRiderNoFrameskip-v4#hyperparameters **in the model card, you have the hyperparameters of the trained agent.**

Reviewer: ```suggestion
If you want to try, check https://huggingface.co/sb3/dqn-BeamRiderNoFrameskip-v4#hyperparameters. There, **in the model card, you have the hyperparameters of the trained agent.**
```

> But finding hyperparameters can be a daunting task. Fortunately, we'll see in the next Unit, how we can **use Optuna for optimizing the Hyperparameters 🔥.**

Reviewer: ```suggestion
But finding hyperparameters can be a daunting task. Fortunately, we'll see in the next bonus Unit, how we can **use Optuna for optimizing the Hyperparameters 🔥.**
```

## Some additional challenges 🏆

The best way to learn **is to try things by your own**!

> In the [Leaderboard](https://huggingface.co/spaces/chrisjay/Deep-Reinforcement-Learning-Leaderboard) you will find your agents. Can you get to the top?

Reviewer: This link is wrong

Author: ✔️
